### DISCUSSION TITLE
Set max dispense limit on dispensers

### DISCUSSION THREAD

**Discussion by jdogresorg** [CounterpartyXCP/CIPs#112] [CIPs]
Currently anyone is able to create dispensers and sell tokens with no limitation on how many dispenses a single dispenser can have. This has led to the creation of dispensers which will stay open and generate millions of dispense transactions and bloat the counterparty database.

I propose that we set a `MAX_DISPENSES` to 1000 and start automatically closing dispensers when they hit or exceed this `MAX_DISPENSES` limit.

A quick check of all counterparty dispensers shows that all but 4 dispensers have less than 1000 dispenses, so 1000 seems like a reasonable limit.

```
mysql> select count(*) as cnt from dispenses group by dispenser_tx_hash_id order by cnt DESC limit 10;
+-------+
| cnt   |
+-------+
| 12273 |
|  8231 |
|  1384 |
|  1047 |
|   965 |
|   603 |
|   333 |
|   187 |
|   186 |
+-------+
10 rows in set (0.06 sec)
```

This change would automatically close dispensers after 1000 dispenses are reached. 

Thoughts?

**Comment by ghost** [comment]
Is there any other elegant way? I would like to pay some fee to maintain my dispenser work forever:

1-1000 dispenses ---->fees take 1% without any payment in advance 

1000-10000 dispenses-----> fees =200 USD

10000 above ----->fees =500 USD

Any ideas?

**Reply by jdogresorg** [reply]
I think this is the most elegant way to address this problem and put some sane limitations on dispensers.

If your dispenser closes, you can simply open another one and get another 1000 dispenses out of it.

IMO setting up some sort of a new dispenser fee structure adds additional costs to using CP dispenser,  complicates the implementation, and doesnt really address the core issue which is database bloat/spam.

**Comment by davestaxcp** [comment]
Not a bad idea. Just trying to think of a counter argument or some sort of circumstance this would be hard to deal with... 

If theoretically someone was to be frustrated by a "limited amount of dispenses" - they could simply change their 'give amount' per dispense to compensate for the 1000 dispense limit.

Might not be a bad idea to post some sort of information or limitation in the code itself for how many "total" dispenses can be created. So in theory you would be restricted to even creating a dispenser that could theoretically dispense over 1000 times.

Another question I would have related to the above is if I had an asset MELONCOCO in a dispenser (1000 total, 1 per each 0.0001 btc sent to dispenser). If someone sent enough btc for 1 total - it counts as "1" dispense... but if someone does one transaction but buys "50" in one transaction, does that count as 51 dispenses, or 2 dispenses?

**Reply by jdogresorg** [reply]
> If theoretically someone was to be frustrated by a "limited amount of dispenses" - they could simply change their 'give amount' per dispense to compensate for the 1000 dispense limit.

To be clear, there is no limitation or change on how you price your items or how many tokens can be sold from your dispenser. You can still sell 10,000,000,000 tokens from a dispenser for whatever amount you want. I am only talking about limiting the total number of dispense transactions a single dispenser can perform before closing.

> Might not be a bad idea to post some sort of information or limitation in the code itself for how many "total" dispenses can be created. So in theory you would be restricted to even creating a dispenser that could theoretically dispense 1000 times.

The CP API could definitely be updated to indicate how many dispenses a dispenser has left, and XChain could display that data as well, so it is clear to all how many dispenses a dispenser has remaining. 

>Another question I would have related to the above is if I had an asset MELONCOCO in a dispenser (1000 total, 1 per each 0.0001 btc sent to dispenser). If someone sent enough btc for 1 total - it counts as "1" dispense... but if someone does one transaction but buys "50" in one transaction, does that count as 51 dispenses, or 2 dispenses?

one BTC payment triggers one dispense... the amount dispensed does not matter...  What matters is the total number of dispense transactions, not the total amount dispensed.

**Reply by davestaxcp** [reply]
good to know!

**Comment by Jpja** [comment]
Good proposal. Nice how the number 1000 is backed by empirics.

Would it be better to enforce this at dispenser *creation*? If a dispenser is set up but >1000 potential dispenses, the client will prevent you from generating the tx.

A counterargument could be that a limit is not necessary if `EMPTY` is redefined to mean no BTC history. https://github.com/CounterpartyXCP/cips/discussions/103#discussioncomment-6703017

Do we need both of these solutions, or would one of them be sufficient?

**Reply by jdogresorg** [reply]
> Would it be better to enforce this at dispenser creation? If a dispenser is set up but >1000 potential dispenses, the client will prevent you from generating the tx.

We could try to prevent CP API from generating the tx, but someone could still hand-build the tx, so would also need logic to reject dispensers trying to get setup which would go over the `MAX_DISPENSES` threshold, at which point, the user just wasted BTC on miners fees on a tx that will ultimately fail. I think its best to leave the create dispensers logic as it is now, and continue to allow ppl to escrow as much as they want in a dispenser.

> Do we need both of these solutions, or would one of them be sufficient?

1. Redefining `EMPTY` to mean no BTC or XCP history should addresses the problem of opening up dispensers on non-owned addresses (like exchange hot-wallet addresses, etc.)

2. Setting a `MAX_DISPENSES` should addresses the problem with the existing dispensers that are already opened on exchange hot-wallets. 

I believe both solutions are necessary in order to prevent abusive dispenses in the future from being setup, and to address the problem with existing dispensers spamming dispense transactions.

**Comment by jdogresorg** [comment]
With the addition of the `origin` field to dispensers, a user will be able to open / refill / close a dispenser on an "EMPTY" address (no BTC/XCP history).

Some users feel that a `MAX_DISPENSE` limit of 1000 is too low, as they have setup dispensers that could dispense hundreds of thousands of times.

I propose an additional rule that when a dispenser is refilled, the `MAX_DISPENSE` counter is reset to 0. This will allow users who want their dispensers to stay open beyond 1,000 dispensers, to enable that functionality by periodically refiling the dispenser with some amount of the token.

**Comment by adamkrellenstein** [comment]
This was implemented.
